<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="sHDL">
<title>Example of sHDL â€¢ sHDL</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Example of sHDL">
<meta property="og:description" content="sHDL">
<meta property="og:image" content="https://now2014.github.io/sHDL/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">sHDL</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Example.html">Example of sHDL</a>
    <a class="dropdown-item" href="../articles/Annotation.html">Genomic annotations</a>
    <a class="dropdown-item" href="../articles/Rebuild.html">Mismatched LD reference panel</a>
    <a class="dropdown-item" href="../articles/MultiGWAS.html">Run sHDL across a large number of traits</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/now2014/sHDL/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Example of sHDL</h1>
            
            <h4 data-toc-skip class="date">2025-11-04</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/now2014/sHDL/blob/main/vignettes/Example.Rmd" class="external-link"><code>vignettes/Example.Rmd</code></a></small>
      <div class="d-none name"><code>Example.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install the latest version of <code>sHDL</code> package via
Github, run the following code in <code>R</code>:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages('remotes')</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"now2014/sHDL"</span>, ref<span class="op">=</span><span class="st">'main'</span><span class="op">)</span></span></code></pre></div>
<p>The <code>data.table</code> package is also required for this
example:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'data.table'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="data-preparation">Data preparation<a class="anchor" aria-label="anchor" href="#data-preparation"></a>
</h2>
<div class="section level3">
<h3 id="pre-computed-ld-reference-panels">Pre-computed LD reference panels<a class="anchor" aria-label="anchor" href="#pre-computed-ld-reference-panels"></a>
</h3>
<p>There are <strong>three</strong> publically available pre-computed
<strong>autosome</strong> LD reference panels for the
<strong>European-ancestry</strong> population (<strong>335,265 Genomic
British individuals in UK Biobank</strong>) from <code>HDL</code>:</p>
<p><strong>Panel 1.</strong> QCed UK Biobank Axiom Array (307,519
SNPs):</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> <span class="at">-t</span> 1 <span class="dt">\</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>  https://www.dropbox.com/s/fuvpwsf6r8tjd6c/UKB_array_SVD_eigen90_extraction.tar.gz<span class="pp">?</span>dl=0 <span class="dt">\</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>  <span class="at">--no-check-certificate</span> <span class="at">-O</span> /path/to/UKB_array_SVD_eigen90_extraction.tar.gz</span></code></pre></div>
<p><strong>Panel 2.</strong> QCed UK Biobank imputed HapMap2 (769,306
SNPs):</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> <span class="at">-t</span> 1 <span class="dt">\</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>  https://www.dropbox.com/s/4vuktycxz1an6sp/UKB_imputed_hapmap2_SVD_eigen99_extraction.tar.gz<span class="pp">?</span>dl=0 <span class="dt">\</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>  <span class="at">--no-check-certificate</span> <span class="at">-O</span> /path/to/UKB_imputed_hapmap2_SVD_eigen99_extraction.tar.gz</span></code></pre></div>
<p><strong>Panel 3.</strong> QCed UK Biobank imputed HapMap3 (1,029,876
SNPs):</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> <span class="at">-t</span> 1 <span class="dt">\</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>  https://www.dropbox.com/s/6js1dzy4tkc3gac/UKB_imputed_SVD_eigen99_extraction.tar.gz<span class="pp">?</span>dl=0 <span class="dt">\</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>  <span class="at">--no-check-certificate</span> <span class="at">-O</span> /path/to/UKB_imputed_SVD_eigen99_extraction.tar.gz</span></code></pre></div>
<p>You may download one of the above reference panel according to the
SNP density of your GWAS summary statistics, or <a href="https://github.com/zhenin/HDL/wiki/Build-a-reference-panel" class="external-link">build
your own LD reference panel</a>. Alternatively, you can download the
reference panels via Baidu Netdisk referring to this <a href="https://github.com/zhenin/HDL/wiki/FAQ" class="external-link">FAQ</a>.</p>
<p>Here, we illustrate the usage of sHDL with the <strong>Panel 2. QCed
UK Biobank Axiom Array</strong>.</p>
<ul>
<li>
<p>Step 01. Download the LD reference panel.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> <span class="at">-t</span> 1 https://www.dropbox.com/s/fuvpwsf6r8tjd6c/UKB_array_SVD_eigen90_extraction.tar.gz<span class="pp">?</span>dl=0 <span class="dt">\ </span>--no-check-certificate <span class="at">-O</span> /your/path/to/UKB_array_SVD_eigen90_extraction.tar.gz</span></code></pre></div>
</li>
<li>
<p>Step 02. Extract files from the archive.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># checking the md5 = ff3fadd7ea08bd29759b6c652618cd1f</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="fu">md5sum</span> /your/path/to/UKB_array_SVD_eigen90_extraction.tar.gz</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="co"># extraction</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="fu">tar</span> <span class="at">-xvf</span> /your/path/to/UKB_array_SVD_eigen90_extraction.tar.gz</span></code></pre></div>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="gwas-summary-statistics">GWAS summary statistics<a class="anchor" aria-label="anchor" href="#gwas-summary-statistics"></a>
</h3>
<p>You can test with the dataset from <code>HDL</code> directly:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">gwas1.example</span>, package<span class="op">=</span><span class="st">'HDL'</span><span class="op">)</span> <span class="co"># GWAS summary statistics from the Neale Lab round 2 GWAS of UK Biobank of birth weight.</span></span>
<span><span class="va">gwas.df</span> <span class="op">&lt;-</span> <span class="va">gwas1.example</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SNP'</span>, <span class="st">'A1'</span>, <span class="st">'A2'</span>, <span class="st">'Z'</span>, <span class="st">'N'</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">gwas1.example</span>, file<span class="op">=</span><span class="st">'gwas.df.rds'</span><span class="op">)</span> <span class="co"># save the gwas.df for sHDL</span></span></code></pre></div>
<p>Or you can prepare your data as follows:</p>
<ul>
<li>
<p>Step 01. Download the Height GWAS summary statistics from the
GIANT consortium.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> https://portals.broadinstitute.org/collaboration/giant/images/f/f7/GIANT_HEIGHT_YENGO_2022_GWAS_SUMMARY_STATS_EUR.gz</span></code></pre></div>
</li>
<li>
<p>Step 02. Load the GWAS summary statistics into
<code>R</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas.df</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">'/path/to/GIANT_HEIGHT_YENGO_2022_GWAS_SUMMARY_STATS_EUR.gz'</span>, fill<span class="op">=</span><span class="cn">TRUE</span>, header<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gwas.df</span>, <span class="fl">5</span><span class="op">)</span> <span class="co"># show the first five rows</span></span>
<span><span class="co">#           SNPID       RSID CHR    POS EFFECT_ALLELE OTHER_ALLELE EFFECT_ALLELE_FREQ         BETA      SE        P       N</span></span>
<span><span class="co"># 1: 1:566875:C:T  rs2185539   1 566875             T            C           0.002800 -0.046315600 0.03930 0.238128  537968</span></span>
<span><span class="co"># 2: 1:728951:C:T rs11240767   1 728951             T            C           0.000356  0.167358000 0.12600 0.185025   85591</span></span>
<span><span class="co"># 3: 1:734462:A:G rs12564807   1 734462             A            G           0.893000  0.004656900 0.01100 0.672866  112953</span></span>
<span><span class="co"># 4: 1:752721:A:G  rs3131972   1 752721             G            A           0.840000  0.000544089 0.00284  0.84811  615932</span></span>
<span><span class="co"># 5: 1:754182:A:G  rs3131969   1 754182             G            A           0.865000  0.001333110 0.00185 0.470389 1100634</span></span></code></pre></div>
</li>
<li>
<p>Step 03. Select &amp; rename <strong>4 required columns</strong>
for <code>sHDL</code>:</p>
<ul>
<li><strong>SNP: rsid of variants</strong></li>
<li><strong>A1: effect allele</strong></li>
<li><strong>A2: other allele</strong></li>
<li><strong>Z: GWAS Z-score</strong></li>
<li><strong>N: Sample size</strong></li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas.df</span><span class="op">$</span><span class="va">Z</span> <span class="op">&lt;-</span> <span class="va">gwas.df</span><span class="op">$</span><span class="va">BETA</span> <span class="op">/</span> <span class="va">gwas.df</span><span class="op">$</span><span class="va">SE</span> <span class="co"># calculate the GWAS Z-score</span></span>
<span><span class="va">gwas.df</span> <span class="op">&lt;-</span> <span class="va">gwas.df</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'RSID'</span>, <span class="st">'EFFECT_ALLELE'</span>, <span class="st">'OTHER_ALLELE'</span>, <span class="st">'Z'</span>, <span class="st">'N'</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">gwas.df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SNP'</span>, <span class="st">'A1'</span>, <span class="st">'A2'</span>, <span class="st">'Z'</span>, <span class="st">'N'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gwas.df</span>, <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#           SNP A1 A2          Z       N</span></span>
<span><span class="co"># 1:  rs2185539  T  C -1.1785140  537968</span></span>
<span><span class="co"># 2: rs11240767  T  C  1.3282381   85591</span></span>
<span><span class="co"># 3: rs12564807  A  G  0.4233545  112953</span></span>
<span><span class="co"># 4:  rs3131972  G  A  0.1915806  615932</span></span>
<span><span class="co"># 5:  rs3131969  G  A  0.7206000 1100634</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">gwas.df</span>, file<span class="op">=</span><span class="st">'gwas.df.rds'</span><span class="op">)</span> <span class="co"># save the gwas.df for sHDL</span></span></code></pre></div>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="genomic-annotation">Genomic annotation<a class="anchor" aria-label="anchor" href="#genomic-annotation"></a>
</h3>
<p>In this example, we will create a binary annotation of all genes
based on the gene locations (hg19).</p>
<ul>
<li>
<p>Step 01. Download the gene location information from
<code>GeneBreak</code> package.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> https://raw.githubusercontent.com/stefvanlieshout/GeneBreak/master/data/ens.gene.ann.hg19.rda</span></code></pre></div>
</li>
<li>
<p>Step 02. Create a binary annotation of all genes.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">'ens.gene.ann.hg19.rda'</span><span class="op">)</span> <span class="co"># load ens.gene.ann.hg19</span></span>
<span><span class="va">LD.path</span> <span class="op">&lt;-</span> <span class="st">'/your/path/to/UKB_array_SVD_eigen90_extraction'</span></span>
<span></span>
<span><span class="co">## extract gene locations ##</span></span>
<span><span class="va">gene.pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">ens.gene.ann.hg19</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">gene.pos</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'chrom'</span>, <span class="st">'start'</span>, <span class="st">'end'</span><span class="op">)</span></span>
<span><span class="va">autosomes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span></span>
<span><span class="va">gene.pos</span> <span class="op">&lt;-</span> <span class="va">gene.pos</span><span class="op">[</span><span class="va">gene.pos</span><span class="op">$</span><span class="va">chrom</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">autosomes</span>, <span class="op">]</span></span>
<span><span class="va">gene.pos</span><span class="op">$</span><span class="va">chrom</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">gene.pos</span><span class="op">$</span><span class="va">chrom</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## extract variants position from LD reference panel ##</span></span>
<span><span class="va">bim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span></span>
<span>  <span class="fu">sHDL</span><span class="fu">:::</span><span class="fu">list.LD.ref.files</span><span class="op">(</span><span class="va">LD.path</span>, suffix<span class="op">=</span><span class="st">'.bim'</span><span class="op">)</span>, <span class="fu">data.table</span><span class="fu">::</span><span class="va"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span>, header<span class="op">=</span><span class="cn">FALSE</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">bim</span> <span class="op">&lt;-</span> <span class="va">bim</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">4</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">bim</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SNP'</span>, <span class="st">'chrom'</span>, <span class="st">'pos'</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## match annotated variants ##</span></span>
<span><span class="va">gene2snp</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">gene.pos</span>, <span class="va">bim</span>, <span class="va">mc.cores</span><span class="op">=</span><span class="fl">4</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">map.gene</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">i</span>, <span class="va">gene.pos.chr</span>, <span class="va">bim.chr</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">s</span> <span class="op">&lt;-</span> <span class="va">gene.pos.chr</span><span class="op">[</span><span class="va">i</span>, <span class="st">'start'</span><span class="op">]</span></span>
<span>    <span class="va">e</span> <span class="op">&lt;-</span> <span class="va">gene.pos.chr</span><span class="op">[</span><span class="va">i</span>, <span class="st">'end'</span><span class="op">]</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">bim.chr</span><span class="op">[</span><span class="va">bim.chr</span><span class="op">$</span><span class="va">pos</span> <span class="op">&gt;=</span> <span class="va">s</span> <span class="op">&amp;</span> <span class="va">bim.chr</span><span class="op">$</span><span class="va">pos</span> <span class="op">&lt;=</span> <span class="va">e</span>, <span class="op">]</span><span class="op">$</span><span class="va">SNP</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">chroms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">bim</span><span class="op">$</span><span class="va">chrom</span><span class="op">)</span></span>
<span>  <span class="va">all.snps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="kw">for</span><span class="op">(</span><span class="va">cur.chrom</span> <span class="kw">in</span> <span class="va">chroms</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">gene.pos.chr</span> <span class="op">&lt;-</span> <span class="va">gene.pos</span><span class="op">[</span><span class="va">gene.pos</span><span class="op">$</span><span class="va">chrom</span> <span class="op">==</span> <span class="va">cur.chrom</span>, <span class="op">]</span></span>
<span>    <span class="va">bim.chr</span> <span class="op">&lt;-</span> <span class="va">bim</span><span class="op">[</span><span class="va">bim</span><span class="op">$</span><span class="va">chrom</span> <span class="op">==</span> <span class="va">cur.chrom</span>, <span class="op">]</span></span>
<span>    <span class="va">snps</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/mclapply.html" class="external-link">mclapply</a></span><span class="op">(</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">gene.pos.chr</span><span class="op">)</span><span class="op">)</span>, <span class="va">map.gene</span>,</span>
<span>      gene.pos.chr<span class="op">=</span><span class="va">gene.pos.chr</span>, bim.chr<span class="op">=</span><span class="va">bim.chr</span>,</span>
<span>      mc.cores<span class="op">=</span><span class="va">mc.cores</span></span>
<span>    <span class="op">)</span></span>
<span>    <span class="va">snps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">snps</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">all.snps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">all.snps</span>, <span class="va">snps</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">all.snps</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">snps</span> <span class="op">&lt;-</span> <span class="fu">gene2snp</span><span class="op">(</span><span class="va">gene.pos</span>, <span class="va">bim</span>, <span class="fl">4</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## create the annotation weights for sHDL ##</span></span>
<span><span class="va">D</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span>, nrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">snps</span><span class="op">)</span>, ncol<span class="op">=</span><span class="fl">1</span>, dimnames<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">snps</span>, <span class="st">'ENS.gene'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">D</span>, file<span class="op">=</span><span class="st">'D.rds'</span><span class="op">)</span> <span class="co"># save the annotation weights for sHDL</span></span></code></pre></div>
</li>
</ul>
<p><strong>Notice</strong>: <code>sHDL</code> will automatically align
<code>D</code> to the LD reference panel according to <code>rsid</code>,
and set the weights of missing variants in the LD reference panel to
<strong>ZERO</strong>.</p>
</div>
</div>
<div class="section level2">
<h2 id="run-shdl">Run sHDL<a class="anchor" aria-label="anchor" href="#run-shdl"></a>
</h2>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/now2014/sHDL" class="external-link">sHDL</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">LD.path</span> <span class="op">&lt;-</span> <span class="st">'/your/path/to/UKB_array_SVD_eigen90_extraction'</span></span>
<span><span class="va">gwas.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'gwas.df.rds'</span><span class="op">)</span></span>
<span><span class="va">D</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'D.rds'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">t0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">sHDL</span><span class="fu">::</span><span class="fu"><a href="../reference/sHDL.html">sHDL</a></span><span class="op">(</span><span class="va">D</span>, <span class="va">gwas.df</span>, <span class="va">LD.path</span>, mc.cores<span class="op">=</span><span class="fl">4</span>, stepwise<span class="op">=</span><span class="cn">TRUE</span>, lam.cut<span class="op">=</span><span class="fl">1</span>, mode<span class="op">=</span><span class="st">'memory'</span>, Dr.path<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="va">t0</span><span class="op">)</span></span></code></pre></div>
<p>sHDL will finish in <strong>~2 mins</strong> with 4 CPU cores.</p>
<p>Down below message in the console during the analysis (birth weight
from <code>HDL</code>) :</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="ex">Starting</span> sHDL analysis...</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="ex">307519</span> out of 307519 <span class="er">(</span><span class="ex">100.00%</span><span class="kw">)</span> <span class="ex">SNPs</span> in reference panel are available in the GWAS.</span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a><span class="ex">Transfoming</span> z <span class="er">(</span><span class="ex">D</span><span class="kw">)</span> <span class="ex">to</span> zr <span class="er">(</span><span class="ex">Dr</span><span class="kw">)</span><span class="ex">...</span></span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a><span class="ex">D</span> is not NULL and Dr.path is not provided. The default path ./sHDL_Dr will be used.</span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a><span class="ex">Annotation</span> name<span class="er">(</span><span class="ex">s</span><span class="kw">)</span><span class="bu">:</span> <span class="pp">[</span><span class="ss">ENS.gene</span><span class="pp">]</span>.</span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a><span class="ex">Applied</span> <span class="kw">`</span><span class="ex">minmax</span><span class="kw">`</span> weight nomalization on [147849 <span class="er">(</span><span class="ex">48.078%</span><span class="kw">)</span><span class="ex">]</span> annotated variants. The theoretical upper boundary for enrichment fold is M / Md = <span class="pp">[</span><span class="ss">2.080</span><span class="pp">]</span>.</span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a><span class="ex">Dr</span> saved to </span>
<span id="cb15-8"><a href="#cb15-8" tabindex="-1"></a><span class="ex">Starting</span> optimization...</span>
<span id="cb15-9"><a href="#cb15-9" tabindex="-1"></a><span class="ex">Optimization</span> done in 112.997 seconds.</span>
<span id="cb15-10"><a href="#cb15-10" tabindex="-1"></a><span class="ex">item</span>    estimation      se      p       note</span>
<span id="cb15-11"><a href="#cb15-11" tabindex="-1"></a><span class="bu">time</span>    112.996711730957        NA      NA      seconds</span>
<span id="cb15-12"><a href="#cb15-12" tabindex="-1"></a><span class="ex">h2</span>      0.357640057353687       0.0752782467546387      2.02504822923168e-06    total heritability</span>
<span id="cb15-13"><a href="#cb15-13" tabindex="-1"></a><span class="ex">intercept</span>       0.920875135658489       0.0241090513453767      3.25119898245832e-319   intercept</span>
<span id="cb15-14"><a href="#cb15-14" tabindex="-1"></a><span class="ex">eta</span>     <span class="at">-0.177230823347607</span>      0.039603506139107       7.63649707679419e-06    eta</span>
<span id="cb15-15"><a href="#cb15-15" tabindex="-1"></a><span class="ex">fold.ENS.gene</span>   1.1880681869999 0.0304909124069927      6.9155713083808e-10     enrichment fold</span>
<span id="cb15-16"><a href="#cb15-16" tabindex="-1"></a><span class="ex">converged</span>       NA      NA      NA      TRUE</span>
<span id="cb15-17"><a href="#cb15-17" tabindex="-1"></a><span class="ex">message</span> NA      NA      NA      CONVERGENCE: REL_REDUCTION_OF_F <span class="op">&lt;</span>= FACTR<span class="pp">*</span>EPSMCH</span>
<span id="cb15-18"><a href="#cb15-18" tabindex="-1"></a><span class="ex">Time</span> difference of 2.824235 mins</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="results-of-shdl">Results of sHDL<a class="anchor" aria-label="anchor" href="#results-of-shdl"></a>
</h2>
<table class="table">
<colgroup>
<col width="15%">
<col width="13%">
<col width="10%">
<col width="8%">
<col width="52%">
</colgroup>
<thead><tr class="header">
<th align="left">item</th>
<th align="right">estimation</th>
<th align="right">se</th>
<th align="right">p</th>
<th align="left">note</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">time</td>
<td align="right">112.9967117</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">seconds</td>
</tr>
<tr class="even">
<td align="left">h2</td>
<td align="right">0.3576401</td>
<td align="right">0.0752782</td>
<td align="right">2.0e-06</td>
<td align="left">total heritability</td>
</tr>
<tr class="odd">
<td align="left">intercept</td>
<td align="right">0.9208751</td>
<td align="right">0.0241091</td>
<td align="right">0.0e+00</td>
<td align="left">intercept</td>
</tr>
<tr class="even">
<td align="left">eta</td>
<td align="right">-0.1772308</td>
<td align="right">0.0396035</td>
<td align="right">7.6e-06</td>
<td align="left">eta</td>
</tr>
<tr class="odd">
<td align="left">fold.ENS.gene</td>
<td align="right">1.1880682</td>
<td align="right">0.0304909</td>
<td align="right">0.0e+00</td>
<td align="left">enrichment fold</td>
</tr>
<tr class="even">
<td align="left">converged</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td align="left">message</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">CONVERGENCE: REL_REDUCTION_OF_F &lt;= FACTR*EPSMCH</td>
</tr>
</tbody>
</table>
<p><strong>Notice</strong>:</p>
<ul>
<li>The item <code>time</code> only indicates the time for maximizing
log likelihood, <strong>NOT</strong> including the time for converting z
(D) to zr (Dr).</li>
<li>The <strong>null hypothesis</strong> of <code>p</code> value for the
fold parameter is <span class="math inline">\(fold \ne 1\)</span>,
<strong>NOT</strong> <span class="math inline">\(fold &gt;
1\)</span>.</li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Ao Lan, Xia Shen.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
