<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Perform heritability enrichment analysis of trait based on GWAS summary statistics."><title>Perform heritability enrichment analysis of trait based on GWAS summary statistics. — sHDL • sHDL</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Perform heritability enrichment analysis of trait based on GWAS summary statistics. — sHDL"><meta property="og:description" content="Perform heritability enrichment analysis of trait based on GWAS summary statistics."><meta property="og:image" content="https://now2014.github.io/sHDL/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">sHDL</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Example.html">Example of sHDL</a>
    <a class="dropdown-item" href="../articles/Annotation.html">Genomic annotations</a>
    <a class="dropdown-item" href="../articles/Rebuild.html">Mismatched LD reference panel</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/now2014/sHDL/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Perform heritability enrichment analysis of trait based on GWAS summary statistics.</h1>
      <small class="dont-index">Source: <a href="https://github.com/now2014/sHDL/blob/main/R/sHDL.R" class="external-link"><code>R/sHDL.R</code></a></small>
      <div class="d-none name"><code>sHDL.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Perform heritability enrichment analysis of trait based on GWAS summary statistics.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">sHDL</span><span class="op">(</span></span>
<span>  <span class="va">D</span>,</span>
<span>  <span class="va">gwas.df</span>,</span>
<span>  <span class="va">LD.path</span>,</span>
<span>  output.file <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nthreads <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  stepwise <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  fill.missing.N <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  lim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="op">-</span><span class="fl">18</span><span class="op">)</span>,</span>
<span>  lam.cut <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  Dr.path <span class="op">=</span> <span class="st">"./Dr"</span>,</span>
<span>  overwrite <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  fix.h2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fix.intercept <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  maxit <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  pgtol <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  start.v <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0.1</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  mode <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"disk"</span>, <span class="st">"memory"</span><span class="op">)</span>,</span>
<span>  pattern <span class="op">=</span> <span class="st">".*chr(\\d{1,2})\\.(\\d{1,2})[_\\.].*"</span>,</span>
<span>  norm.method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"minmax"</span>, <span class="st">"scaled"</span>, <span class="st">"none"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>D</dt>
<dd><p>A vector of genomic annotations with vector names of SNP IDs.</p></dd>


<dt>gwas.df</dt>
<dd><p>A data frame including GWAS summary statistics of genetic variants for a trait.
The input data frame should include following columns: SNP, SNP ID; A1, effect allele; A2, reference allele;
N, sample size; Z, z-score; If Z is not given, alternatively, you may provide: b, estimate of marginal effect in GWAS; se, standard error of the estimates of marginal effects in GWAS.</p></dd>


<dt>LD.path</dt>
<dd><p>Path to the directory where linkage disequilibrium (LD) information is stored, compatible with HDL LD reference panels.</p></dd>


<dt>output.file</dt>
<dd><p>Where the log and results should be written. If you do not specify a file, the log will be printed on the console.</p></dd>


<dt>nthreads</dt>
<dd><p>Number of threads to use, default <code>nthreads = 1</code>.</p></dd>


<dt>stepwise</dt>
<dd><p>Whether to estimate enrichment fold by estimating heritability and intercept first, default <code>stepwise = FALSE</code>. If <code>fix.h2</code> and <code>fix.intercept</code> are not <code>NULL</code>, <code>stepwise</code> will be overridden.</p></dd>


<dt>fill.missing.N</dt>
<dd><p>If the sample size is missing in the GWAS summary statistics, <code>fill.missing.N</code> is used to fill the missing values.</p></dd>


<dt>lim</dt>
<dd><p>Tolerance limitation to ensure the positive-definiteness of covariance matrices, default <code>lim = exp(-18)</code>.</p></dd>


<dt>lam.cut</dt>
<dd><p>Eigenvalue cutoff for LD matrices, default <code>lam.cut = NULL</code>, which means no cutoff.</p></dd>


<dt>Dr.path</dt>
<dd><p>Path to the directory where the Dr matrices are stored, default <code>Dr.path = "./Dr"</code>.</p></dd>


<dt>overwrite</dt>
<dd><p>Whether to overwrite the existing Dr matrices, default <code>overwrite = FALSE</code>.</p></dd>


<dt>verbose</dt>
<dd><p>Whether to print the log on the console, default <code>verbose = FALSE</code>.</p></dd>


<dt>fix.h2</dt>
<dd><p>Whether to fix the heritability to <code>fix.h2</code> or estimate the heritability, default <code>fix.h2 = NULL</code>, which means estimate the heritability.</p></dd>


<dt>fix.intercept</dt>
<dd><p>Whether to fix the intercept to <code>fix.intercept</code> or estimate the intercept, default <code>fix.intercept = NULL</code>, which means estimate the intercept.</p></dd>


<dt>maxit</dt>
<dd><p>Maximum number of iterations, default <code>maxit = 1000</code>.</p></dd>


<dt>pgtol</dt>
<dd><p>Tolerance for convergence, default <code>pgtol = 1e-3</code>.</p></dd>


<dt>start.v</dt>
<dd><p>A vector of starting values <code>c(fold, h2, intercept)</code> for optimization.</p></dd>


<dt>mode</dt>
<dd><p>Whether to store Dr to disk or memory, default <code>mode = "disk"</code>. If <code>mode = "disk"</code>, <code>Dr</code> is stored to disk (path returned only) and lam are not returned. If <code>mode = "memory"</code>, <code>Dr</code> and <code>lam</code> are returned.</p></dd>


<dt>pattern</dt>
<dd><p>Chromosome and picece pattern of LD files, default is <code>".*chr(\d{1,2})\.(\d{1,2})[_\.].*"</code>.</p></dd>


<dt>norm.method</dt>
<dd><p>The normalization method, either <code>"minmax"</code> (default), <code>"scaled"</code> or <code>"none"</code>. If <code>"minmax"</code>, the annotation weight vector <code>D</code> is normalized to [0, 1]. If <code>"scaled"</code>, the sum of normalized vector <code>D</code> is scaled to the number of annotated SNPs. If <code>"none"</code>, the annotation weight vector <code>D</code> is not normalized.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A list is returned with:</p><ul><li><p>time The time used for optimization.</p></li>
<li><p>fold The estimated heritability enrichment fold.</p></li>
<li><p>h2 The estimated SNP-based heritability.</p></li>
<li><p>intercept The estimated intercept.</p></li>
<li><p>fold.se The standard error of the estimated heritability enrichment fold.</p></li>
<li><p>h2.se The standard error of the estimated heritability.</p></li>
<li><p>intercept.se The standard error of the estimated intercept.</p></li>
<li><p>fold.p P-value based on Wald test for the estimated heritability enrichment fold.</p></li>
<li><p>h2.p P-value based on Wald test for the estimated heritability.</p></li>
<li><p>intercept.p P-value based on Wald test for the estimated intercept.</p></li>
<li><p>stepwise Whether the optimization is done in a stepwise manner.</p></li>
<li><p>converged Whether the optimization converges.</p></li>
<li><p>message The message returned by <code><a href="https://rdrr.io/r/stats/optim.html" class="external-link">optim</a></code>.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="note">Note<a class="anchor" aria-label="anchor" href="#note"></a></h2>
    <p>Users can download the precomputed eigenvalues and eigenvectors of LD correlation matrices for European ancestry population. The download link can be found at https://github.com/zhenin/HDL/wiki/Reference-panels
These are the LD matrices and their eigen-decomposition from 335,265 genomic British UK Biobank individuals. Three sets of reference panel are provided:
1) 1,029,876 QCed UK Biobank imputed HapMap3 SNPs. The size is about 33 GB after unzipping. Although it takes more time, using the imputed panel provides more accurate estimates of genetic correlations.
Therefore if the GWAS includes most of the HapMap3 SNPs, then we recommend using the imputed reference panel.
2) 769,306 QCed UK Biobank imputed HapMap2 SNPs. The size is about 18 GB after unzipping.If one of your GWAS includes most of the HapMap 2 SNPs, but many SNPs (more than 1<!-- %) in the above HapMap 3 reference panel are absent, -->
then this HapMap2 panel is more proper to be used for HDL.
3) 307,519 QCed UK Biobank Axiom Array SNPs. The size is about 7.5 GB after unzipping.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Lan A and Shen X (2024). Modeling the Genetic Architecture of Complex Traits via Stratified High-Definition Likelihood.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Ao Lan, Xia Shen</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co">## The GWAS summary statistics for birth weight loaded from HDL package.</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">gwas1.example</span>, package<span class="op">=</span><span class="st">'HDL'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">M</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">gwas1.example</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">D</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html" class="external-link">rbinom</a></span><span class="op">(</span><span class="va">M</span>, <span class="fl">1</span>, <span class="fl">0.01</span><span class="op">)</span> <span class="co"># random D vector</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">D</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">gwas1.example</span><span class="op">$</span><span class="va">SNP</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## The path to the directory where linkage disequilibrium (LD) information is stored.</span></span></span>
<span class="r-in"><span><span class="va">LD.path</span> <span class="op">&lt;-</span> <span class="st">"path/to/UKB_array_SVD_eigen90_extraction"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## To speed up the test, we set a large lam.cut value.</span></span></span>
<span class="r-in"><span><span class="va">res.sHDL</span> <span class="op">&lt;-</span> <span class="fu">sHDL</span><span class="op">(</span><span class="va">D</span>, <span class="va">gwas1.example</span>, <span class="va">LD.path</span>, nthreads<span class="op">=</span><span class="fl">4</span>, stepwise<span class="op">=</span><span class="cn">TRUE</span>, lam.cut<span class="op">=</span><span class="fl">10</span>, Dr.path<span class="op">=</span><span class="cn">NULL</span>, mode<span class="op">=</span><span class="st">"memory"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">res.sHDL</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Ao Lan, Xia Shen.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

