<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Maximize the log likelihood to infer the heritability enrichment folds jointly."><title>Maximize the log likelihood to infer the heritability enrichment folds jointly. — sHDL.joint • sHDL</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Maximize the log likelihood to infer the heritability enrichment folds jointly. — sHDL.joint"><meta property="og:description" content="Maximize the log likelihood to infer the heritability enrichment folds jointly."><meta property="og:image" content="https://now2014.github.io/sHDL/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">sHDL</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Example.html">Example of sHDL</a>
    <a class="dropdown-item" href="../articles/Annotation.html">Genomic annotations</a>
    <a class="dropdown-item" href="../articles/Rebuild.html">Mismatched LD reference panel</a>
    <a class="dropdown-item" href="../articles/MultiGWAS.html">Run sHDL across a large number of traits</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/now2014/sHDL/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Maximize the log likelihood to infer the heritability enrichment folds jointly.</h1>
      <small class="dont-index">Source: <a href="https://github.com/now2014/sHDL/blob/main/R/sHDL.joint.R" class="external-link"><code>R/sHDL.joint.R</code></a></small>
      <div class="d-none name"><code>sHDL.joint.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Maximize the log likelihood to infer the heritability enrichment folds jointly.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">sHDL.joint</span><span class="op">(</span></span>
<span>  <span class="va">Dr.path</span>,</span>
<span>  <span class="va">gwas.df</span>,</span>
<span>  <span class="va">LD.path</span>,</span>
<span>  output.file <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  annots <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  mc.cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  stepwise <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  fill.missing.N <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"none"</span>, <span class="st">"min"</span>, <span class="st">"max"</span>, <span class="st">"median"</span>, <span class="st">"mean"</span><span class="op">)</span>,</span>
<span>  lim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="op">-</span><span class="fl">18</span><span class="op">)</span>,</span>
<span>  lam.cut <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  fix.h2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fix.intercept <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fix.eta <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  maxit <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  pgtol <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  start.v <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  lwr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  upr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pattern <span class="op">=</span> <span class="st">".*chr(\\d{1,2})\\.(\\d{1,2})[_\\.].*"</span>,</span>
<span>  par.h2 <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  nthreads <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>Dr.path</dt>
<dd><p>The path to the pre-computed <code>diag(Dr)</code> data, see also <code>sHDL::sHDL.joint.reduct.dim</code>.</p></dd>


<dt>gwas.df</dt>
<dd><p>A data frame including GWAS summary statistics of genetic variants for a trait.
The input data frame should include following columns: SNP, SNP ID; A1, effect allele; A2, reference allele;
N, sample size; Z, z-score; If Z is not given, alternatively, you may provide: b, estimate of marginal effect in GWAS; se, standard error of the estimates of marginal effects in GWAS.</p></dd>


<dt>LD.path</dt>
<dd><p>Path to the directory where linkage disequilibrium (LD) information is stored, compatible with HDL LD reference panels.</p></dd>


<dt>output.file</dt>
<dd><p>Where the log and results should be written. If you do not specify a file, the log will be printed on the console.</p></dd>


<dt>annots</dt>
<dd><p>A vector of annotation names, default <code>NULL</code>, which means all annotations in the <code>Dr.path</code> will be used.</p></dd>


<dt>mc.cores</dt>
<dd><p>Number of cores to use for parallelization, default <code>mc.cores = 1</code>.</p></dd>


<dt>stepwise</dt>
<dd><p>Whether to estimate enrichment fold by estimating heritability, intercept and eta first, default <code>stepwise = FALSE</code>. If <code>fix.h2</code> and <code>fix.intercept</code> are not <code>NULL</code>, <code>stepwise</code> will be overridden.</p></dd>


<dt>fill.missing.N</dt>
<dd><p>If the sample size is missing in the GWAS summary statistics, <code>fill.missing.N</code> is used to fill the missing values.</p></dd>


<dt>lim</dt>
<dd><p>Tolerance limitation to ensure the positive-definiteness of covariance matrices, default <code>lim = exp(-18)</code>.</p></dd>


<dt>lam.cut</dt>
<dd><p>Eigenvalue cutoff for LD matrices, default <code>lam.cut = NULL</code>, which means no cutoff. For analyses with a limited number of traits and annotations, a lower cutoff (such as 0.1, or even not using a cutoff at all) is recommended. For large-scale analyses, a higher cutoff (such as 1)  is recommended, to yield fast computation.</p></dd>


<dt>verbose</dt>
<dd><p>Whether to print the log on the console, default <code>verbose = FALSE</code>.</p></dd>


<dt>fix.h2</dt>
<dd><p>Whether to fix the heritability to <code>fix.h2</code> or estimate the heritability, default <code>fix.h2 = NULL</code>, which means estimate the heritability.</p></dd>


<dt>fix.intercept</dt>
<dd><p>Whether to fix the intercept to <code>fix.intercept</code> or estimate the intercept, default <code>fix.intercept = NULL</code>, which means estimate the intercept.</p></dd>


<dt>fix.eta</dt>
<dd><p>Whether to fix the eta to <code>fix.eta</code> or estimate the eta, default <code>fix.eta = NULL</code>, which means estimate the eta. If <code>fix.eta</code> is not <code>NULL</code>, the eta will be fixed to <code>fix.eta</code>.</p></dd>


<dt>maxit</dt>
<dd><p>Maximum number of iterations, default <code>maxit = 1000</code>.</p></dd>


<dt>pgtol</dt>
<dd><p>Tolerance for convergence, default <code>pgtol = 1e-3</code>.</p></dd>


<dt>start.v</dt>
<dd><p>Starting values for <code>c(h2, intercept, eta, fold1, fold2, ..., foldn)</code>, where n is the number of annotations. Default is <code>NULL</code>, which means <code>c(0.1, 1, 0, 1, ..., 1)</code>. If <code>fix.h2</code>, <code>fix.intercept</code> or <code>fix.eta</code> is not <code>NULL</code>, the starting values for <code>h2</code>, <code>intercept</code> or <code>eta</code> will be ignored.</p></dd>


<dt>lwr</dt>
<dd><p>Lower bounds for <code>c(h2, intercept, eta, fold1, fold2, ..., foldn)</code>. Default is <code>NULL</code>, which means <code>c(0, 0, -2, 0, 0, ..., 0)</code>. When <code>fix.h2</code>, <code>fix.intercept</code> or <code>fix.eta</code> is not <code>NULL</code>, the lower bounds for <code>h2</code>, <code>intercept</code> or <code>eta</code> will be ignored.</p></dd>


<dt>upr</dt>
<dd><p>Upper bounds for <code>c(h2, intercept, eta, fold1, fold2, ..., foldn)</code>, default is <code>NULL</code>, which means <code>c(1, 5, 1, M/Md_1, M/Md_2, ..., M/Md_n)</code>, where <code>Md_i</code> is the sum of the weights of the i-th annotation. When <code>fix.h2</code>, <code>fix.intercept</code> or <code>fix.eta</code> is not <code>NULL</code>, the upper bounds for <code>h2</code>, <code>intercept</code> or <code>eta</code> will be ignored.</p></dd>


<dt>pattern</dt>
<dd><p>Chromosome and picece pattern of LD files, default is <code>".*chr(\\d{1,2})\\.(\\d{1,2})[_\\.].*"</code>.</p></dd>


<dt>par.h2</dt>
<dd><p>Whether to estimate the partitioned heritability, default <code>par.h2 = FALSE</code>.</p></dd>


<dt>nthreads</dt>
<dd><p>Number of threads to use for matrix operations, default <code>nthreads = 1</code>. The default value is suitable for most cases, do not change it unless you are sure about the performance.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A data.frame is returned with:</p><ul><li><p>item  The name of the parameter.</p></li>
<li><p>estimation  The estimated value of the parameter.</p></li>
<li><p>se  The standard error of the parameter.</p></li>
<li><p>p  The p-value of the parameter.</p></li>
<li><p>note  The note of the parameter.</p></li>
</ul></div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Ao Lan, Xia Shen.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

